helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacity.java

```diff
@@ -42,79 +49,102 @@ public class WagedInstanceCapacity implements InstanceCapacityDataProvider {
+      Map<String, Resource> resourceMap, WagedResourceWeightsProvider weightProvider) {
+    processPendingMessages(cache, currentStateOutput, resourceMap, weightProvider);
+    processCurrentState(cache, currentStateOutput, resourceMap, weightProvider);
   }
 
   /**
    * Process the pending messages based on the Current states
    * @param currentState - Current state of the resources.
    */
-  public void processPendingMessages(CurrentStateOutput currentState) {
-    Map<String, Map<Partition, Map<String, Message>>> pendingMsgs = currentState.getPendingMessages();
-
-    for (String resource : pendingMsgs.keySet()) {
-      Map<Partition, Map<String, Message>> partitionMsgs = pendingMsgs.get(resource);
+  public void processPendingMessages(ResourceControllerDataProvider cache,
+      CurrentStateOutput currentState, Map<String, Resource> resourceMap,
+      WagedResourceWeightsProvider weightProvider) {
+
+    for (Map.Entry<String, Resource> resourceEntry : resourceMap.entrySet()) {
+      String resName = resourceEntry.getKey();
+      Resource resource = resourceEntry.getValue();
+      // list of partitions in the resource
+      Collection<Partition> partitions = resource.getPartitions();
+      // State model definition for the resource
+      StateModelDefinition stateModelDef = cache.getStateModelDef(resource.getStateModelDefRef());
+      if (stateModelDef == null) {
+        LOG.warn("State Model Definition for resource: " + resName + " is null");
+        continue;
+      }
+      Map<String, Integer> statePriorityMap = stateModelDef.getStatePriorityMap();
 
-      for (Partition partition : partitionMsgs.keySet()) {
+      for (Partition partition : partitions) {
         String partitionName = partition.getPartitionName();
-
         // Get Partition Weight
-        Map<String, Integer> partCapacity = getPartitionCapacity(resource, partitionName);
-
-        // TODO - check
-        Map<String, Message> msgs = partitionMsgs.get(partition);
-        // TODO - Check
-        for (String instance : msgs.keySet()) {
-           reduceAvailableInstanceCapacity(instance, partCapacity);
+        Map<String, Integer> partCapacity = weightProvider.getPartitionWeights(resName, partitionName);
+
+        // Get the pending messages for the partition
+        Map<String, Message> pendingMessages = currentState.getPendingMessageMap(resName, partition);
+        if (pendingMessages != null && !pendingMessages.isEmpty()) {
+          for (Map.Entry<String, Message> entry :  pendingMessages.entrySet()) {
+            String instance = entry.getKey();
+            if (isPartitionInAllocatedMap(instance, resName, partitionName)) {
+              continue;
+            }
+            Message msg = entry.getValue();
+            if (statePriorityMap.get(msg.getFromState()) < statePriorityMap.get(msg.getToState())
+                && msg.getToState().equals(stateModelDef.getInitialState())
```

Comment:
I cannot remember but not sure whether this can cause double counting. Because when there is a OFFLINE -> STANDBY message, then the current state is already OFFLINE. So the capacity will be reduced by pending message and current states both.

Because I remember the order of creating state transition with state model with OFFLINE state. But I cannot remember when we write it back to ZK...

Reply:
I was worried about double-counting as a result, we maintain 
private final Map<String, Map<String, Set<String>>> _allocatedPartitionsMap;
As long as for same instance, same resource/partition, if we have either pending message and/or currentState, we will account for only once as we maintain this additional metadata.